### ISSUE TITLE
Anima pipeline via API produces solid green output — `sd_model_checkpoint` change does not load `forge_additional_modules`

### ISSUE THREAD

**Issue by V0aRyn** [Haoming02/sd-webui-forge-classic#1291]
---

**Describe the bug**

When an Anima checkpoint is selected and generated through the **API** (`/sdapi/v1/txt2img`), the output is a solid green/low-entropy image (~20KB for 768×1152). The same model works correctly when selected and generated from the **GUI**.

**Root cause (two independent issues)**

*Issue 1 — Model path not scanned*

When Forge is started **without** `--forge-ref-comfy-home`, the `models/diffusion_models/` directory is never added to the checkpoint scan paths. Anima models placed there won't appear in `sdapi/v1/sd-models`, and setting `sd_model_checkpoint` to their filename returns `"Model not found..."`.

The scanning happens in `modules/launch_utils.py:configure_comfy_reference()` which only runs when `--forge-ref-comfy-home` is set. Adding `--ckpt-dirs "path/to/models/diffusion_models"` manually resolves this.

*Issue 2 — API bypasses `modules_change`, so Qwen VAE + text encoder are never loaded*

When `sd_model_checkpoint` is changed via `/sdapi/v1/options`, `modules/ui.py:156–157` returns `gr.skip()` for this key, short-circuiting the normal options handler. This means `refresh_model_loading_parameters()` (in `modules_forge/main_entry.py:118`) is **never called**, so `model_data.forge_loading_parameters["additional_modules"]` stays empty.

When the model is subsequently loaded (`modules/sd_models.py:338`), it reads:
```python
additional_state_dicts = model_data.forge_loading_parameters.get("additional_modules", [])
```
which is empty. The Anima pipeline loads without `qwen_image_vae.safetensors` or `qwen_3_06b_base.safetensors`. The transformer runs inference correctly, but decoding through the VAE produces garbage (solid green output).

Through the GUI, switching presets calls `modules_change()` (`modules_forge/main_entry.py:170–187`) which saves `forge_additional_modules` to `shared.opts`, then calls `refresh_model_loading_parameters()` — so the modules are loaded correctly.

**To reproduce**

```bash
# 1. Start Forge without --forge-ref-comfy-home
python launch.py --api --port 7869

# 2. Place anima-preview3-base.safetensors in models/diffusion_models/

# 3. Switch preset and checkpoint via API
curl -s -X POST http://127.0.0.1:7869/sdapi/v1/options \
  -H "Content-Type: application/json" \
  -d '{
    "forge_preset": "anima",
    "sd_vae": "qwen_image_vae.safetensors",
    "sd_model_checkpoint": "anima-preview3-base.safetensors"
  }'

# 4. Generate
curl -s -X POST http://127.0.0.1:7869/sdapi/v1/txt2img \
  -H "Content-Type: application/json" \
  -d '{
    "prompt": "masterpiece, best quality, score_7, safe, 1girl",
    "negative_prompt": "worst quality, low quality, score_1, score_2, score_3",
    "steps": 32,
    "sampler_name": "Euler a",
    "cfg_scale": 4,
    "width": 768,
    "height": 1152
  }'

# Result: ~20KB image, visually solid green
```

**Workaround (confirmed)**

1. Start Forge with `--ckpt-dirs "D:\path\to\models\diffusion_models"` so the Anima model is discoverable.
2. Open the GUI once, switch UI Preset to `anima` — this triggers `modules_change()` which populates `forge_additional_modules` and calls `refresh_model_loading_parameters()`.
3. Subsequent API calls to `/sdapi/v1/txt2img` will then work correctly (the modules are cached in `model_data.forge_loading_parameters`).

Alternatively, call the equivalent of `modules_change(module_values, preset="anima", save=True, refresh=True)` programmatically before generating.

**Expected behavior**

Changing `sd_model_checkpoint` via API should also load the corresponding `forge_additional_modules` for the active preset, or at least trigger `refresh_model_loading_parameters()` so the modules are picked up — just as they are when the same change is made through the GUI.

**Suggested fix**

A minimal fix in `modules/ui.py:156–157`: instead of unconditionally returning `gr.skip()` for `"sd_model_checkpoint"` and `"forge_additional_modules"`, allow the call-through when the change is made via the API (non-Gradio context), so `refresh_model_loading_parameters()` runs.

A more complete fix would add a dedicated API endpoint `/sdapi/v1/switch-preset` that wraps `preset_change()` + `modules_change()` + `refresh_model_loading_parameters()` in the correct order.

**References**

- Related Anima feature discussion: #671
- Related Qwen model path setup: #553
- `modules_forge/main_entry.py` — `modules_change()` at line 165, `refresh_model_loading_parameters()` at line 118
- `modules/sd_models.py` — `forge_model_reload()` at line 310, reads `additional_modules` at line 338

**Environment**

- Forge branch: neo (Haoming02/sd-webui-forge-classic)
- Python: 3.11.9 (Windows)
- torch: 2.11.0+cu130
- GPU: NVIDIA RTX 4070 SUPER (12GB)
- Model tested: `anima-preview3-base.safetensors` and `anima-base-v1.0.safetensors`
- Issue also affects Qwen-Image models loaded via API

---

*This issue was researched and authored with assistance from Hermes Agent (Nous Research). The reproduction, root-cause analysis, and suggested fix were empirically verified.*

**Comment by V0aRyn** [comment]
**Update: Fixed!** 🎉

Root cause confirmed: `sysinfo.py:set_config()` processed `forge_preset` as a plain option via `shared.opts.set()`, so changing the preset via API never triggered `modules_change()` or `checkpoint_change()`. The Qwen VAE + text encoder were never loaded, causing NaN in VAE decoding.

**Fix** (in `modules/sysinfo.py`):
- `forge_preset` changes now apply checkpoint, additional modules, and dtype synchronously — matching the GUI preset switch behavior
- Explicit `sd_model_checkpoint` / `forge_additional_modules` / `forge_unet_storage_dtype` in the same API request are now attributed to the target preset, regardless of JSON key order
- Regression tests added in `tests/test_sysinfo_set_config.py`

**Verified**: API generation now returns 82KB+ valid images (was 20KB green/NaN) with the same `forge_preset=anima` + `sd_model_checkpoint=miaomiaoHarem_anima14` + Qwen VAE/text encoder configuration that previously failed.

**Comment by V0aRyn** [comment]
**Final resolution — follow-up finding** 🎯

After verifying the fix above, one more critical detail surfaced:

**The Anima pipeline requires  to be explicitly passed in API calls.**

Without it, the VAE decoder still produces NaN even with the correct modules loaded. The default scheduler used by the API endpoint is not compatible with Anima's Cosmos-based architecture.

Working API call (minimal reproduction):


The GUI works without this because Gradio's preset change handler already sets the correct scheduler from `forge_preset_t2i_scheduler`. The API endpoint doesn't.

**Complete fix checklist:**
1. ✅ `modules/sysinfo.py` — `forge_preset` changes now trigger `modules_change` + `checkpoint_change` + `dtype_change` synchronously (thanks Codex)
2. ✅ Regression tests in `tests/test_sysinfo_set_config.py`
3. 📝 **API callers must include `"scheduler": "Normal"`** — this is a documentation/API-contract requirement
4. 📝 Tall portrait resolutions (768×1152) are unstable; prefer square (1024×1024) or near-square

The 20KB green/NaN issue is fully resolved under these conditions.

**Comment by Haoming02** [comment]
1. So... you created a folder yourself, then complain that the folder is not scanned?
2. Usually, you just use `override_settings` to specify the `sd_model_checkpoint` and `forge_additional_modules` ; alternatively, apparently the `/sdapi/v1/options` api also calls the `set_config` function in `sysinfo.py`, so you can use that too
